### PR TITLE
feat: User Profile Properties can't edit

### PR DESCRIPTION
**Pull request by jakiuncle**
feat: User Profile Properties can't edit

### CODE REVIEW

**Comment by jakiuncle** [comment]
draft

**Comment by hsluoyz** [comment]
@leo220yuyaodog plz review

File: web/src/common/PropertyTable.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+              <Button style={{marginRight: 8}} onClick={() => this.save(table, index)}>
+              Save
+              </Button>
+              <Popconfirm title="Sure to cancel?" onConfirm={() => this.cancel(table, index)}>
+                <a>Cancel</a>
+              </Popconfirm>
+            </span>
+          ) : (
+            <div>
+              <Tooltip placement="topLeft" title="Edit">
```
Comment (leo220yuyaodog) [inline web/src/common/PropertyTable.js:189]:
```suggestion
              <Tooltip placement="topLeft" title={i18next.t("general:Edit")}>
```

File: web/src/common/PropertyTable.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+                <Button disabled={this.state.editingIndex !== ""} style={{marginRight: "5px"}} icon={<EditOutlined />} size="small" onClick={() => this.edit(record, index)} />
+              </Tooltip>
+              <Tooltip placement="topLeft" title="Delete">
```
Comment (leo220yuyaodog) [inline web/src/common/PropertyTable.js:192]:
```suggestion
              <Tooltip placement="topLeft" title={i18next.t("general:Delete")}>
```

**Review by leo220yuyaodog — COMMENTED** [review]
Lgtm. Can add some i18 translate.

**Review by leo220yuyaodog — APPROVED** [review]

**Comment by hsluoyz** [comment]
@leo220yuyaodog

**Comment by hsluoyz** [comment]
@jakiuncle provide a demo video

**Comment by jakiuncle** [comment]
https://user-images.githubusercontent.com/88994283/208352868-7a03fb35-b7e5-4bba-9c09-ea1c5bf00068.MP4

**Comment by jakiuncle** [comment]
@hsluoyz plz review

**Comment by jakiuncle** [comment]
https://user-images.githubusercontent.com/88994283/208886063-81c16bce-bbfd-401b-90de-3f7fe86e2f37.mp4

File: web/src/UserEditPage.js [hunk 3]
```diff
@@ -28,8 +28,8 @@ import SamlWidget from "./common/SamlWidget";
 import SelectRegionBox from "./SelectRegionBox";
 import WebAuthnCredentialTable from "./WebauthnCredentialTable";
 import ManagedAccountTable from "./ManagedAccountTable";
+import PropertyTable from "./common/PropertyTable";
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:31]: import PropertyTable from "./PropertyTable";

File: web/src/UserEditPage.js [hunk 4]
```diff
@@ -28,8 +28,8 @@ import SamlWidget from "./common/SamlWidget";
...
 
-import {Controlled as CodeMirror} from "react-codemirror2";
 import "codemirror/lib/codemirror.css";
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:33]: these 3 lines?

File: web/src/UserEditPage.js [hunk 5]
```diff
@@ -116,7 +116,15 @@ class UserEditPage extends React.Component {
     // if ([].includes(key)) {
     //   value = Setting.myParseInt(value);
     // }
-    return value;
+    const obj = new Object();
+    if (["properties"].includes(key)) {
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:120]: What's this?

File: web/src/common/PropertyTable.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React from "react";
+import {DeleteOutlined} from "@ant-design/icons";
+import {Button, Input, Table, Tooltip} from "antd";
+import * as Setting from "../Setting";
+import i18next from "i18next";
+
+class PolicyTable extends React.Component {
```
Comment (hsluoyz) [inline web/src/common/PropertyTable.js:21]: PropertyTable

File: web/src/common/PropertyTable.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  constructor(props) {
+    super(props);
+    const keys = Object.keys(this.props.properties);
+    const properties = [];
+    for (let i = 0; i < keys.length; i++) {
+      const property = new Object();
```
Comment (hsluoyz) [inline web/src/common/PropertyTable.js:27]: ?

File: web/src/common/PropertyTable.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  updateField(table, index, key, value) {
+    table[index][key] = value;
+    this.updateTable(table);
+  }
+
+  addRow(table) {
+    const row = {index: crypto.randomUUID()};
+    table = Setting.addRow(table, row, "top");
```
Comment (hsluoyz) [inline web/src/common/PropertyTable.js:51]: top?

File: web/src/common/PropertyTable.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+  deleteRow(table, i) {
+    table = Setting.deleteRow(table, i);
+    this.updateTable(table);
+  }
+
+  renderTable(table) {
+    const columns = [
+      {
+        title: "Keys",
```
Comment (hsluoyz) [inline web/src/common/PropertyTable.js:66]: i18n?

File: web/src/common/PropertyTable.js [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        render: (text, record, index) => {
+          return (
+            <Input value={text} onChange={e => {
+              this.updateField(table, index, "key", e.target.value);
+            }} />
+          );
+        },
+      },
+      {
+        title: "Values",
```
Comment (hsluoyz) [inline web/src/common/PropertyTable.js:78]: i18n?

File: web/src/common/PropertyTable.js [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+          <div>
+            <Button style={{marginRight: "5px"}} type="primary" size="small" onClick={() => this.addRow(table)}>{i18next.t("general:Add")}</Button>
+          </div>
+        )}
+      />
+    );
+  }
+
+  render() {
+    return (<>
```
Comment (hsluoyz) [inline web/src/common/PropertyTable.js:119]: <>?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Pull request closed without merging**